src/v/cluster/self_test_frontend.cc

```diff
@@ -0,0 +1,305 @@
+    co_return enrich(
+      node_id(), _self_be.local().get_status_report(get_results));
+}
+
+/// Implementation of methods in self_test_frontend::remote_invoke
+
+ss::future<self_test_frontend::local_test_state>
+self_test_frontend::remote_invoke::start_test(uuid_t uuid) {
+    auto v = co_await send_self_test_rpc(
+      [uuid](self_test_rpc_client_protocol c) {
+          return c.start_test(
+            start_test_request{.id = uuid},
+            rpc::client_opts(raft::clock_type::now() + 5s));
+      });
+    co_return enrich(node_id(), v);
+}
+
+ss::future<self_test_frontend::local_test_state>
+self_test_frontend::remote_invoke::stop_test() {
+    auto v = co_await send_self_test_rpc([](self_test_rpc_client_protocol c) {
+        /// Bigger timeout to wait for executing jobs to shutdown
+        return c.stop_test(
+          stop_test_request{}, rpc::client_opts(raft::clock_type::now() + 30s));
+    });
+    co_return enrich(node_id(), v);
+}
+
+ss::future<self_test_frontend::local_test_state>
+self_test_frontend::remote_invoke::query_status(bool get_results) {
+    auto v = co_await send_self_test_rpc(
+      [get_results](self_test_rpc_client_protocol c) {
+          return c.query_status(
+            query_status_request{.return_results = get_results},
+            rpc::client_opts(raft::clock_type::now() + 5s));
+      });
+    co_return enrich(node_id(), v);
+}
+
+/// Implementation of methods in self_test_frontend
+
+self_test_frontend::self_test_frontend(
+  model::node_id self,
+  ss::sharded<cluster::members_table>& members,
+  ss::sharded<cluster::partition_leaders_table>& leaders,
+  ss::sharded<cluster::self_test_backend>& self_be,
+  ss::sharded<rpc::connection_cache>& connections)
+  : _self(self)
+  , _members(members)
+  , _leaders(leaders)
+  , _self_be(self_be)
+  , _connections(connections) {}
+
+ss::future<> self_test_frontend::start() { return ss::now(); }
+
+ss::future<> self_test_frontend::stop() { co_await _gate.close(); }
+
+ss::future<self_test_frontend::err_or_id_t> self_test_frontend::start_test() {
+    auto gate_holder = _gate.hold();
+
+    auto leader_opt = _leaders.local().get_leader(model::controller_ntp);
+    if (!leader_opt.has_value() || leader_opt != _self) {
+        // Only cluster leader can initiaite start_test() rpc calls
+        co_return self_test_status_report::failure;
+    }
+    const auto st = co_await status();
+    if (st.is_out_of_sync()) {
+        // Result contains inconsistencies, cancel and report why
+        co_await stop_test();
+        co_return self_test_status_report::out_of_sync;
+    } else if (st.is_test_active()) {
+        // no inconsistences, but test active, hasn't finished
+        co_return self_test_status_report::running;
+    }
+
+    /// Invoke command to start test on all nodes, using the same test id
+    const auto id = uuid_t::create();
+    const auto results = co_await invoke_on_all_nodes(
```

Comment:
Bit unclear whether the "results" here are an interim status report, or if the start_test RPC blocks until the test is complete?